src/transformers/models/whisper/generation_whisper.py

```diff
@@ -1016,94 +1072,211 @@ def _set_language_and_task(language, task, is_multilingual, generation_config):
                 )
             generation_config.task = task
 
-    @staticmethod
-    def _set_forced_decoder_ids(task, language, prompt_ids, generation_config, config, kwargs):
-        forced_decoder_ids = None
-        # Legacy code for backward compatibility
-        if hasattr(config, "forced_decoder_ids") and config.forced_decoder_ids is not None:
-            forced_decoder_ids = config.forced_decoder_ids
+    def _retrieve_init_tokens(self, input_features, generation_config, config, num_segment_frames, kwargs):
+        def replace_or_add(lst: List[int], num: int, itr: Iterator[int]):
+            """short function to replace num with a itr in lst"""
+            found = any(i in lst for i in itr)
+            if found:
+                lst = [num if i in itr else i for i in lst]
+            else:
+                lst.append(num)
+            return lst
+
+        if kwargs.get("forced_decoder_ids", None) is not None:
+            forced_decoder_ids = kwargs["forced_decoder_ids"]
         elif hasattr(generation_config, "forced_decoder_ids") and generation_config.forced_decoder_ids is not None:
             forced_decoder_ids = generation_config.forced_decoder_ids
+        elif hasattr(config, "forced_decoder_ids") and config.forced_decoder_ids is not None:
```

Comment:
Good with me to prioritise the generation config over the config (and as discussed offline at https://huggingface.slack.com/archives/C02G13FEMDH/p1706544739238239?thread_ts=1706447656.541159&cid=C02G13FEMDH)